File: apps/src/lib/api/model-router.ts
import type {
  LatestSessionModelApplyResult,
  ModelRouteBindingListResult,
  ModelRouteBindingSavePayload,
  ModelRouteBindingSaveResult,
  ModelRouteQuickCheckResult,
  ModelRouterImportResult,
  ProbeApplySelectedPayload,
  ProbeRunAllResult,
  ProbeRunListResult,
  ProbeRunSummary,
  SessionModelListResult,
  SessionModelSummary,
  SessionModelUpdateResult,
  WorkspaceModelDefaultSummary,
} from "@/types/model-router";
import { invoke, withAddr } from "./transport";

export const modelRouterClient = {
  async listSessions(workspace?: string): Promise<SessionModelListResult> {
    return invoke<SessionModelListResult>(
      "service_model_router_session_list",
      withAddr({ workspace: workspace || null }),
    );
  },

  async updateSessionModel(params: {
    threadId: string;
    model: string;
    reasoningEffort?: string | null;
    source?: string | null;
    locked?: boolean;
  }): Promise<SessionModelUpdateResult> {
    return invoke<SessionModelUpdateResult>(
      "service_model_router_session_update_model",
      withAddr({
        threadId: params.threadId,
        model: params.model,
        reasoningEffort: params.reasoningEffort || null,
        source: params.source || "manual",
        locked: typeof params.locked === "boolean" ? params.locked : null,
      }),
    );
  },

  async applyModelToLatestWorkspaceSession(params: {
    workspace: string;
    model: string;
    reasoningEffort?: string | null;
    source?: string | null;
    locked?: boolean;
  }): Promise<LatestSessionModelApplyResult> {
    return invoke<LatestSessionModelApplyResult>(
      "service_model_router_session_apply_latest_for_workspace",
      withAddr({
        workspace: params.workspace,
        model: params.model,
        reasoningEffort: params.reasoningEffort || null,
        source: params.source || "manual",
        locked: typeof params.locked === "boolean" ? params.locked : null,
      }),
    );
  },

  async setSessionSubagentModel(params: {
    parentThreadId: string;
    model: string;
    reasoningEffort?: string | null;
    source?: string | null;
  }): Promise<SessionModelSummary> {
    return invoke<SessionModelSummary>(
      "service_model_router_session_set_subagent_model",
      withAddr({
        parentThreadId: params.parentThreadId,
        model: params.model,
        reasoningEffort: params.reasoningEffort || null,
        source: params.source || "manual",
      }),
    );
  },

  async clearSessionSubagentModel(parentThreadId: string): Promise<boolean> {
    return invoke<boolean>(
      "service_model_router_session_clear_subagent_model",
      withAddr({ parentThreadId }),
    );
  },

  async setWorkspaceDefault(params: {
    workspace: string;
    defaultModel?: string | null;
    defaultReasoningEffort?: string | null;
    inheritLastSession?: boolean;
    autoRemember?: boolean;
  }): Promise<WorkspaceModelDefaultSummary> {
    return invoke<WorkspaceModelDefaultSummary>(
      "service_model_router_workspace_default_set",
      withAddr({
        workspace: params.workspace,
        defaultModel: params.defaultModel || null,
        defaultReasoningEffort: params.defaultReasoningEffort || null,
        inheritLastSession:
          typeof params.inheritLastSession === "boolean"
            ? params.inheritLastSession
            : null,
        autoRemember:
          typeof params.autoRemember === "boolean" ? params.autoRemember : null,
      }),
    );
  },

  deleteWorkspaceDefault(workspace: string): Promise<boolean> {
    return invoke<boolean>(
      "service_model_router_workspace_default_delete",
      withAddr({ workspace }),
    );
  },

  async listBindings(model?: string): Promise<ModelRouteBindingListResult> {
    return invoke<ModelRouteBindingListResult>(
      "service_model_router_binding_list",
      withAddr({ model: model || null }),
    );
  },

  async saveBinding(
    payload: ModelRouteBindingSavePayload,
  ): Promise<ModelRouteBindingSaveResult> {
    return invoke<ModelRouteBindingSaveResult>(
      "service_model_router_binding_save",
      withAddr({
        id: payload.id || null,
        model: payload.model,
        aggregateApiId: payload.aggregateApiId,
        enabled: payload.enabled ?? null,
        priority: payload.priority ?? null,
        weight: payload.weight ?? null,
        routeStrategy: payload.routeStrategy || null,
        manualPreferred: payload.manualPreferred ?? null,
        supportsResponses: payload.supportsResponses ?? null,
        supportsChatCompletions: payload.supportsChatCompletions ?? null,
        requiresAdapter: payload.requiresAdapter ?? null,
      }),
    );
  },

  deleteBinding(id: string): Promise<unknown> {
    return invoke("service_model_router_binding_delete", withAddr({ id }));
  },

  async runProbe(
    aggregateApiId: string,
    options: { signal?: AbortSignal } = {},
  ): Promise<ProbeRunSummary> {
    return invoke<ProbeRunSummary>(
      "service_model_router_probe_run",
      withAddr({ aggregateApiId }),
      options,
    );
  },

  async runAllProbes(options: { signal?: AbortSignal } = {}): Promise<ProbeRunAllResult> {
    return invoke<ProbeRunAllResult>(
      "service_model_router_probe_run_all",
      withAddr({}),
      options,
    );
  },

  async addManualProbeModel(params: {
    aggregateApiId: string;
    model: string;
    supportsResponses?: boolean;
    supportsChatCompletions?: boolean;
    requiresAdapter?: boolean;
  }): Promise<ProbeRunSummary> {
    return invoke<ProbeRunSummary>(
      "service_model_router_probe_manual_model",
      withAddr({
        aggregateApiId: params.aggregateApiId,
        model: params.model,
        supportsResponses:
          typeof params.supportsResponses === "boolean"
            ? params.supportsResponses
            : null,
        supportsChatCompletions:
          typeof params.supportsChatCompletions === "boolean"
            ? params.supportsChatCompletions
            : null,
        requiresAdapter:
          typeof params.requiresAdapter === "boolean" ? params.requiresAdapter : null,
      }),
    );
  },

  async quickCheck(
    params: {
      aggregateApiId: string;
      model: string;
    },
    options: { signal?: AbortSignal } = {},
  ): Promise<ModelRouteQuickCheckResult> {
    return invoke<ModelRouteQuickCheckResult>(
      "service_model_router_probe_quick_call",
      withAddr({
        aggregateApiId: params.aggregateApiId,
        model: params.model,
      }),
      options,
    );
  },

  async applyProbe(probeRunId: string): Promise<ProbeRunSummary> {
    return invoke<ProbeRunSummary>(
      "service_model_router_probe_apply",
      withAddr({ probeRunId }),
    );
  },

  async applySelectedProbeCandidates(
    payload: ProbeApplySelectedPayload,
  ): Promise<ProbeRunSummary> {
    return invoke<ProbeRunSummary>(
      "service_model_router_probe_apply_selected",
      withAddr({
        probeRunId: payload.probeRunId,
        candidateIds: payload.candidateIds,
      }),
    );
  },

  async listProbeRuns(limit = 20): Promise<ProbeRunListResult> {
    return invoke<ProbeRunListResult>(
      "service_model_router_probe_list",
      withAddr({ limit }),
    );
  },

  async importCodexManager(sourcePath?: string | null): Promise<ModelRouterImportResult> {
    return invoke<ModelRouterImportResult>(
      "service_model_router_import_codexmanager",
      withAddr({ sourcePath: sourcePath || null }),
    );
  },
};

File: apps/src-tauri/src/commands/model_router.rs
use crate::commands::shared::rpc_call_in_background;

#[tauri::command]
pub async fn service_model_router_session_list(
    addr: Option<String>,
    workspace: Option<String>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({ "workspace": workspace });
    rpc_call_in_background("modelRouter/session/list", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_session_update_model(
    addr: Option<String>,
    thread_id: String,
    model: String,
    reasoning_effort: Option<String>,
    source: Option<String>,
    locked: Option<bool>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "threadId": thread_id,
        "model": model,
        "reasoningEffort": reasoning_effort,
        "source": source,
        "locked": locked,
    });
    rpc_call_in_background("modelRouter/session/updateModel", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_session_apply_latest_for_workspace(
    addr: Option<String>,
    workspace: String,
    model: String,
    reasoning_effort: Option<String>,
    source: Option<String>,
    locked: Option<bool>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "workspace": workspace,
        "model": model,
        "reasoningEffort": reasoning_effort,
        "source": source,
        "locked": locked,
    });
    rpc_call_in_background("modelRouter/session/applyLatestForWorkspace", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_session_set_subagent_model(
    addr: Option<String>,
    parent_thread_id: String,
    model: String,
    reasoning_effort: Option<String>,
    source: Option<String>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "parentThreadId": parent_thread_id,
        "model": model,
        "reasoningEffort": reasoning_effort,
        "source": source,
    });
    rpc_call_in_background("modelRouter/session/subagentModel/set", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_session_clear_subagent_model(
    addr: Option<String>,
    parent_thread_id: String,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "parentThreadId": parent_thread_id,
    });
    rpc_call_in_background("modelRouter/session/subagentModel/clear", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_workspace_default_set(
    addr: Option<String>,
    workspace: String,
    default_model: Option<String>,
    default_reasoning_effort: Option<String>,
    inherit_last_session: Option<bool>,
    auto_remember: Option<bool>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "workspace": workspace,
        "defaultModel": default_model,
        "defaultReasoningEffort": default_reasoning_effort,
        "inheritLastSession": inherit_last_session,
        "autoRemember": auto_remember,
    });
    rpc_call_in_background("modelRouter/workspaceDefault/set", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_workspace_default_delete(
    addr: Option<String>,
    workspace: String,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "workspace": workspace,
    });
    rpc_call_in_background("modelRouter/workspaceDefault/delete", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_binding_list(
    addr: Option<String>,
    model: Option<String>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({ "model": model });
    rpc_call_in_background("modelRouter/binding/list", addr, Some(params)).await
}

#[allow(clippy::too_many_arguments)]
#[tauri::command]
pub async fn service_model_router_binding_save(
    addr: Option<String>,
    id: Option<String>,
    model: String,
    aggregate_api_id: String,
    enabled: Option<bool>,
    priority: Option<i64>,
    weight: Option<i64>,
    route_strategy: Option<String>,
    manual_preferred: Option<bool>,
    supports_responses: Option<bool>,
    supports_chat_completions: Option<bool>,
    requires_adapter: Option<bool>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "id": id,
        "model": model,
        "aggregateApiId": aggregate_api_id,
        "enabled": enabled,
        "priority": priority,
        "weight": weight,
        "routeStrategy": route_strategy,
        "manualPreferred": manual_preferred,
        "supportsResponses": supports_responses,
        "supportsChatCompletions": supports_chat_completions,
        "requiresAdapter": requires_adapter,
    });
    rpc_call_in_background("modelRouter/binding/save", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_binding_delete(
    addr: Option<String>,
    id: String,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({ "id": id });
    rpc_call_in_background("modelRouter/binding/delete", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_probe_run(
    addr: Option<String>,
    aggregate_api_id: String,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({ "aggregateApiId": aggregate_api_id });
    rpc_call_in_background("modelRouter/probe/run", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_probe_run_all(
    addr: Option<String>,
) -> Result<serde_json::Value, String> {
    rpc_call_in_background("modelRouter/probe/runAll", addr, None).await
}

#[tauri::command]
pub async fn service_model_router_probe_manual_model(
    addr: Option<String>,
    aggregate_api_id: String,
    model: String,
    supports_responses: Option<bool>,
    supports_chat_completions: Option<bool>,
    requires_adapter: Option<bool>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "aggregateApiId": aggregate_api_id,
        "model": model,
        "supportsResponses": supports_responses,
        "supportsChatCompletions": supports_chat_completions,
        "requiresAdapter": requires_adapter,
    });
    rpc_call_in_background("modelRouter/probe/manualModel", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_probe_quick_call(
    addr: Option<String>,
    aggregate_api_id: String,
    model: String,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "aggregateApiId": aggregate_api_id,
        "model": model,
    });
    rpc_call_in_background("modelRouter/probe/quickCall", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_probe_apply(
    addr: Option<String>,
    probe_run_id: String,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({ "probeRunId": probe_run_id });
    rpc_call_in_background("modelRouter/probe/apply", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_probe_list(
    addr: Option<String>,
    limit: Option<i64>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({ "limit": limit });
    rpc_call_in_background("modelRouter/probe/list", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_model_router_import_codexmanager(
    addr: Option<String>,
    source_path: Option<String>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({ "sourcePath": source_path });
    rpc_call_in_background("modelRouter/import/codexManager", addr, Some(params)).await
}
